bounded/src/set.rs

```diff
@@ -511,332 +608,267 @@ fn nullary_id_to_app(term: Term, relations: &[RelationDecl]) -> Term {
-        match self {
-            Ast::Insert(set, element) => Ok(Update {
-                index: indices[&(set.as_str(), element)],
-                value: true,
-            }),
-            Ast::Remove(set, element) => Ok(Update {
-                index: indices[&(set.as_str(), element)],
-                value: false,
-            }),
-            _ => Err(TranslationError::ExpectedUpdate(self)),
+        Term::Quantified {
+            quantifier: Quantifier::Exists,
+            binders,
+            body,
+        } => {
+            let vecs = binders
+                .iter()
+                .map(|b| cardinality(context.universe, &b.sort))
+                .map(|card| (0..card).collect::<Vec<Element>>())
+                .multi_cartesian_product_fixed()
+                .map(|elements| {
+                    let mut new_assignments = assignments.clone();
+                    assert_eq!(binders.len(), elements.len());
+                    for (binder, element) in binders.iter().zip(elements) {
+                        new_assignments.insert(binder.name.to_string(), element);
+                    }
+                    traverse_disjunction(body, context, &new_assignments, func)
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            vecs.into_iter().flatten().collect()
         }
-    }
-
-    fn contradicts(&self, other: &Ast) -> bool {
-        matches!((self, other),
-            (Ast::Includes(a, b), Ast::Excludes(c, d))
-            | (Ast::Excludes(a, b), Ast::Includes(c, d))
-            | (Ast::Insert(a, b), Ast::Remove(c, d))
-            | (Ast::Insert(a, b), Ast::NoOp(c, d))
-            | (Ast::Remove(a, b), Ast::Insert(c, d))
-            | (Ast::Remove(a, b), Ast::NoOp(c, d))
-            | (Ast::NoOp(a, b), Ast::Insert(c, d))
-            | (Ast::NoOp(a, b), Ast::Remove(c, d))
-                if a == c && b == d
-        )
-    }
+        term => vec![func(term, assignments)?],
+    };
+    Ok(vec)
 }
 
-fn and(terms: BTreeSet<Ast>) -> Ast {
-    // flatten
-    let terms: BTreeSet<Ast> = terms.into_iter().flat_map(Ast::get_and).collect();
-    // short circuit if possible
-    for a in &terms {
-        if *a == Ast::Or(btreeset![]) || terms.iter().any(|b| a.contradicts(b)) {
-            return Ast::Or(btreeset![]);
-        }
-    }
-    // remove redundant updates
-    let old = terms;
-    let mut terms = btreeset![];
-    for term in &old {
-        if let Ast::Insert(set, element) = term.clone() {
-            if old.contains(&Ast::Includes(set.clone(), element.clone())) {
-                terms.insert(Ast::NoOp(set, element));
-                continue;
-            }
-        } else if let Ast::Remove(set, element) = term.clone() {
-            if old.contains(&Ast::Excludes(set.clone(), element.clone())) {
-                terms.insert(Ast::NoOp(set, element));
-                continue;
-            }
+fn term_to_transition(
```

Comment:
Can you put a comment on this function as well? It's also big. Can you try to describe the transformation it does at a high level and in particular what input terms will succeed/fail on?

Also took me a while to figure out that `traverse_disjunction` is the "outer" function and `term_to_transition` is the "inner" function.